feat: Extract minimal DI integration to OpenFeature.Providers.DependencyInjection

arttonoyan opened the pull request:
> [!IMPORTANT] 
**Update. 2025-11-10**
Following the discussion in [https://github.com/open-feature/dotnet-sdk/pull/596#discussion_r2495975595](https://github.com/open-feature/dotnet-sdk/pull/596#discussion_r2495975595). the package name in this proposal is updated from `OpenFeature.DependencyInjection` to `OpenFeature.Providers.DependencyInjection`. Rationale. clearer scope for provider authors. thinner provider packages. reduced coupling with the SDK.

## This PR
Extract the minimal, provider-agnostic DI surface into a new package: `OpenFeature.DependencyInjection.Abstractions`.
This isolates the contracts and lightweight wiring needed to integrate any OpenFeature provider without pulling in a concrete implementation.

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

Fixes #587 

### Notes
* **Options split:** Extracted provider-agnostic configuration from `OpenFeatureOptions` into a new base options type: **`OpenFeatureProviderOptions`**.
* **Inheritance:** `OpenFeatureOptions` now **inherits** from `OpenFeatureProviderOptions`.
* **Internal wiring update:** Internal configuration that previously targeted `OpenFeatureOptions` now targets `OpenFeatureProviderOptions`.

### Before (internal)

```csharp
services.AddOptions<OpenFeatureOptions>()
    .Configure(options =>
    {
        options.AddProviderName(null);
    });
```

### After (internal)

```csharp
services.AddOptions<OpenFeatureProviderOptions>()
    .Configure(options =>
    {
        options.AddProviderName(null);
    });
```

> **Note:** This is an internal refactor. It **does not affect consumers** unless they were directly modifying `OpenFeatureOptions` via DI options configuration (which is not a supported/typical usage).

---

### Impact

* **No new functionality.**
* **Behavior unchanged** at runtime (same provider selection, same evaluation outcomes, same logs).
* **Binary/source compatibility:** Consumers using the documented extension methods and standard registration remain unaffected.
* **Potential touchpoints:** Only internal code paths (or unconventional consumer code directly configuring `OpenFeatureOptions`) require updates to target `OpenFeatureProviderOptions`.

---

### How to test

This is a **regression-only** refactor.

**Expectations**

* All existing unit/integration tests pass **unchanged**.
* No DI resolution or startup configuration differences.

arttonoyan committed:
Add OpenFeature.DependencyInjection.Abstractions project

Introduced a new project, `OpenFeature.DependencyInjection.Abstractions`, to the solution. This project supports dependency injection abstractions and targets multiple frameworks (`netstandard2.0`, `net8.0`, `net9.0`, `net462`) for broad compatibility.

Configured the project with the `Microsoft.NET.Sdk` SDK and set the root namespace to `OpenFeature.DependencyInjection.Abstractions`. Added dependencies on `Microsoft.Extensions.DependencyInjection.Abstractions` and `Microsoft.Extensions.Options` to enable DI and options configuration.

arttonoyan committed:
Refactor OpenFeature for DI and provider abstraction

Refactored the OpenFeature framework to introduce `OpenFeatureProviderBuilder`, enhancing support for dependency injection and provider management.

- Changed namespaces to align with DI abstractions.
- Made `FeatureCodes` public for broader accessibility.
- Added `InternalsVisibleTo` for testing and project references.
- Introduced `OpenFeatureProviderBuilder` for managing providers and policies.
- Added extension methods for provider and policy registration.
- Refactored `OpenFeatureBuilder` to inherit from `OpenFeatureProviderBuilder`.
- Consolidated shared functionality in `OpenFeatureProviderOptions`.
- Updated `FeatureBuilderExtensions` and `InMemoryProviderOptions` to use the new abstraction.
- Updated tests to reflect new method signatures and hierarchy.
- Removed redundant methods and properties, improving code organization.

These changes improve maintainability, extensibility, and alignment with modern DI patterns.

arttonoyan committed:
Refactor AddProvider API for simplicity and consistency

Simplified the `AddProvider` API by removing the `TFeatureProvider` generic type parameter and directly using the `FeatureProvider` type. Updated the `TOptions` parameter to ensure it derives from `OpenFeatureProviderOptions`.

Added a `<ProjectReference>` to `OpenFeature.csproj` in `OpenFeature.DependencyInjection.Abstractions.csproj`. Updated `OpenFeatureOptions` to `OpenFeatureProviderOptions` in the default name selector policy.

Refactored `AddInMemoryProvider` methods to align with the new API. Updated tests in `OpenFeatureBuilderExtensionsTests` to reflect the changes and validate the updated functionality.

Performed general code cleanup, including XML documentation updates, to improve clarity and maintain consistency across the codebase.

arttonoyan committed:
Refactor FeatureLifecycleManager initialization

Refactored `FeatureLifecycleManager` to modularize provider, hook, and handler initialization with new methods: `InitializeProvidersAsync`, `InitializeHooks`, and `InitializeHandlers`. Updated `EnsureInitializedAsync` to use these methods for improved readability and maintainability.

Revised `AddInMemoryProvider` in `FeatureBuilderExtensions` to use a generic `FeatureProvider` abstraction. Adjusted `CreateProvider` methods accordingly.

Improved code clarity in `FeatureFlagIntegrationTest` by renaming variables for consistency and removing redundant assignments.

arttonoyan committed:
Refactor tests to use OpenFeatureProviderOptions

Updated FeatureLifecycleManagerTests to replace OpenFeatureOptions
with OpenFeatureProviderOptions for configuring feature providers.
Added support for hooks and keyed singletons to enhance modularity.
Introduced additional feature flag retrieval in FeatureFlagIntegrationTest.
Added dependency on OpenFeature.DependencyInjection.Abstractions.

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

kylejuliandev commented on src/OpenFeature.DependencyInjection.Abstractions/OpenFeature.DependencyInjection.Abstractions.csproj:
```diff
@@ -0,0 +1,22 @@
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net8.0;net9.0;net462</TargetFrameworks>
+    <RootNamespace>OpenFeature.DependencyInjection.Abstractions</RootNamespace>
```
suggestion: Should we have these abstractions in their own namespace? Would it be more accessible to have them in say just the `OpenFeature` namespace?

arttonoyan replied:
@kylejuliandev Great question - thanks for calling it out.
I’d keep these in their own namespace for now. `OpenFeature.DependencyInjection.Abstractions` targets provider authors (integrating new providers), not typical app consumers. Keeping it separate avoids cluttering the root `OpenFeature` namespace and lets us evolve DI-specific types independently.
Of course, I’m open to discussion about this.

askpt replied:
I prefer the @arttonoyan. This is a very specific Abstraction layer. 

I am wondering if we should keep in a `OpenFeature.Abstractions` instead. This would be a "library" of just `OpenFeature` related abstractions like "hooks", "provider"... I am happy to keep `OpenFeature.DependencyInjection.Abstractions` if the rest of the team agrees.

arttonoyan replied:
@askpt @kylejuliandev

**Proposal**
Rename the package to `OpenFeature.Providers.DependencyInjection`. This name better reflects scope. It is intuitive for provider authors who want to add DI integration.

**Follow-up idea**
In a separate discussion, consider extracting common provider contracts into `OpenFeature.Providers.Abstractions`. We would then have two clear packages. `OpenFeature.Providers.Abstractions`, and `OpenFeature.Providers.DependencyInjection`.

**Why this helps**

* Clear separation. abstractions live in one place, DI glue in another.
* Thinner provider packages. they depend on `Abstractions`, optionally on the DI helpers.
* Cleaner dependency graph. no DI references inside core contracts.
* Easier discovery on NuGet. consistent `OpenFeature.Providers.*` naming.

cc: @beeme1mr

askpt replied:
@arttonoyan @beeme1mr @kylejuliandev: I think this is a great discussion.

I would propose for 2.9 to create:
- `OpenFeature.Providers.DependencyInjection` (this PR)

For 2.10:
- `OpenFeature.Providers.Abstractions`
- `OpenFeature.Hooks.Abstractions`

The story that is happening in Providers, is also true for the Hooks.

kylejuliandev commented on src/OpenFeature.Hosting/Internal/FeatureLifecycleManager.cs:
```diff
@@ -34,7 +43,12 @@ public async ValueTask EnsureInitializedAsync(CancellationToken cancellationToke
         }
+    }
 
+    /// <inheritdoc />
```
suggestion: Not sure we need the `inheritdoc` section here

kylejuliandev commented on src/OpenFeature.DependencyInjection.Abstractions/OpenFeatureProviderBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+/// <summary>
+/// Describes a <see cref="OpenFeatureProviderBuilder"/> backed by an <see cref="IServiceCollection"/>.
+/// </summary>
+public abstract class OpenFeatureProviderBuilder(IServiceCollection services)
```
Oh wait I see, these are provider specific. Nvm this comment.

kylejuliandev commented on src/OpenFeature.Hosting/OpenFeatureServiceCollectionExtensions.cs:
```diff
@@ -49,7 +50,7 @@ public static IServiceCollection AddOpenFeature(this IServiceCollection services
                 options.DefaultNameSelector = provider =>
                 {
-                    var options = provider.GetRequiredService<IOptions<OpenFeatureOptions>>().Value;
+                    var options = provider.GetRequiredService<IOptions<OpenFeatureProviderOptions>>().Value;
```
suggestion: it should be fine for this to remain `OpenFeatureOptions` right? `OpenFeatureOptions` extends `OpenFeatureProviderOptions`

arttonoyan replied:
@kylejuliandev Fixed: https://github.com/open-feature/dotnet-sdk/pull/596/commits/f52132178ebf244fd8de653cbe5897f42c731c33

arttonoyan committed:
Removev ID property

arttonoyan committed:
Merge branch 'feature/di-abstractions-builder-api-587' of https://github.com/arttonoyan/dotnet-sdk into feature/di-abstractions-builder-api-587

arttonoyan commented on OpenFeature.slnx:
```diff
@@ -48,6 +48,7 @@
     <Project Path="samples/AspNetCore/Samples.AspNetCore.csproj" />
   </Folder>
   <Folder Name="/src/">
+    <Project Path="src/OpenFeature.DependencyInjection.Abstractions/OpenFeature.DependencyInjection.Abstractions.csproj" Id="607a23fa-2c2f-4d2e-9079-adae18326512" />
```
Removed: https://github.com/open-feature/dotnet-sdk/pull/596/commits/3d680f3f054cba3b9547f276d09a8016bf26c66d

arttonoyan commented on src/OpenFeature.DependencyInjection.Abstractions/Diagnostics/FeatureCodes.cs:
```diff
@@ -22,7 +22,7 @@ namespace OpenFeature.Hosting.Diagnostics;
 /// </code>
 /// </example>
-internal static class FeatureCodes
+public static class FeatureCodes
```
@kylejuliandev **Answer:** Yes - within the scope of this PR, `FeatureCodes` are still required; the current changes reference them.
**Follow-up proposal:** It’s a good time to graduate them from `Experimental`. Not in this PR - I suggest a separate task.
@askpt @beeme1mr what do you think about me opening a follow-up issue to:
* remove the `Experimental` attribute,
* update docs/usages,
* add a changelog entry and brief migration note?

askpt reviewed (commented):
Great work! I only added a couple of minor comments. 

Please have a chat with @kylejuliandev to try understand the impact on his deprecations efforts. I remember he mentioned a couple of fixes he was working on for the DI and I don't want that effort to get lost during this PR.

kylejuliandev commented on src/OpenFeature.Hosting/OpenFeatureOptions.cs:
```diff
@@ -1,52 +1,12 @@
-        }
-    }
-
     private readonly HashSet<string> _hookNames = [];
```
I went ahead over the weekend and gave it a stab in https://github.com/open-feature/dotnet-sdk/pull/609. Seems simple enough. I think if we register two hooks of the same type, then we should still handle them both. Maybe let's take this conversation over there. Happy to close this here as it's not really relevant to the changes

kylejuliandev commented on src/OpenFeature.DependencyInjection.Abstractions/OpenFeatureProviderBuilderExtensions.cs:
```diff
@@ -0,0 +1,145 @@
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using OpenFeature.DependencyInjection.Abstractions;
+
+namespace OpenFeature;
```
These extensions are available to devs directly when they setup OpenFeature with the OpenFeature builder, I think leaving them in the `OpenFeature` namespace is less disruptive and makes it more easily accessible

arttonoyan committed:
Update src/OpenFeature.Hosting/Internal/FeatureLifecycleManager.cs

Co-authored-by: André Silva <2493377+askpt@users.noreply.github.com>
Signed-off-by: Artyom Tonoyan <arttonoyan@gmail.com>

arttonoyan committed:
Update src/OpenFeature.Hosting/Internal/FeatureLifecycleManager.cs

Co-authored-by: André Silva <2493377+askpt@users.noreply.github.com>
Signed-off-by: Artyom Tonoyan <arttonoyan@gmail.com>

arttonoyan committed:
Refactor OpenFeatureOptions to OpenFeatureProviderOptions

Refactored `OpenFeatureOptions` to `OpenFeatureProviderOptions`
to improve clarity and maintainability. Updated all references
and test cases to use the new class. Removed inheritance of
`OpenFeatureOptions` from `OpenFeatureProviderOptions` and
introduced `_hookNames` for managing hook names.

Replaced `TestOptions` with `TestProviderOptions` in
`OpenFeatureBuilderExtensionsTests`, adding a `SomeFlag`
property for more flexible testing. Renamed and updated
`OpenFeatureOptionsTests` to `OpenFeatureProviderOptionsTests`.

Performed general cleanup, including removing unused
namespaces, ensuring consistent naming conventions, and
aligning method calls with the new class structure.

askpt commented on src/OpenFeature.DependencyInjection.Abstractions/OpenFeature.DependencyInjection.Abstractions.csproj:
```diff
@@ -0,0 +1,22 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net8.0;net9.0;net462</TargetFrameworks>
```
```suggestion
```

Please remove this. It is now managed centrally by https://github.com/open-feature/dotnet-sdk/pull/611

arttonoyan committed:
Refactor Dependency Injection Abstractions

Migrated `OpenFeature.DependencyInjection.Abstractions` to a new project, `OpenFeature.Providers.DependencyInjection`, to improve modularity and maintainability. Updated namespaces, project references, and `using` directives accordingly.

Removed the deprecated `OpenFeature.DependencyInjection.Abstractions` project and its associated files. Introduced `OpenFeature.Providers.DependencyInjection.csproj` to house the migrated functionality.

Updated test files to reference the new namespace and adjusted experimental feature annotations. Reorganized build configuration files to align with the new project structure. Ensured documentation and metadata reflect the refactored codebase.

arttonoyan committed:
chore: Merge dev into current branch and resolve merge conflicts

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Remove inheritdoc from FeatureLifecycleManager

arttonoyan committed:
Merge branch 'feature/di-abstractions-builder-api-587' of https://github.com/arttonoyan/dotnet-sdk into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Clean up unused namespaces in test files

Removed `OpenFeature.Hosting.Internal` namespace from
`FeatureLifecycleManagerTests.cs` and
`OpenFeatureBuilderExtensionsTests.cs` as it is no longer
required. Adjusted the placement of
`OpenFeature.Providers.DependencyInjection` for consistency
and improved readability. These changes streamline imports
and enhance maintainability.

arttonoyan commented:
@askpt @kylejuliandev Could you please do a final review? I’d like to merge this PR.

arttonoyan committed:
Update release.yml

Signed-off-by: Artyom Tonoyan <arttonoyan@gmail.com>

arttonoyan commented on src/OpenFeature.Providers.DependencyInjection/OpenFeature.Providers.DependencyInjection.csproj:
Will be done within the task https://github.com/open-feature/dotnet-sdk/issues/633

kylejuliandev commented on src/OpenFeature.Providers.DependencyInjection/OpenFeatureProviderBuilderExtensions.cs:
```diff
@@ -0,0 +1,145 @@
+        where TOptions : PolicyNameOptions
+    {
+        if (builder == null) throw new ArgumentNullException(nameof(builder));
+        if (configureOptions == null) throw new ArgumentNullException(nameof(configureOptions));
```
suggestion(non-blocking): on a subsequent PR maybe we should include some sensible error messages when throwing these exceptions

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

arttonoyan committed:
Add cancellation checks and null safety in providers

Introduce cancellation checks in InitializeProvidersAsync to
allow prompt operation cancellation. Add null-coalescing
operator to ensure OpenFeatureProviderOptions are configured,
throwing an InvalidOperationException if not. Improve code
readability by adjusting formatting.

arttonoyan committed:
Cast AddProvider return type to OpenFeatureBuilder

Explicitly cast the return type of AddProvider method calls to
OpenFeatureBuilder in FeatureBuilderExtensions.cs to ensure
type specificity. Updated the ArgumentNullException message
for null 'configure' parameter to provide clearer guidance
on the need for a valid configuration.

askpt commented:
> @askpt @kylejuliandev I ran into a blocking issue. I tried to avoid any major changes, but it looks like I can’t solve it without doing at least some refactoring. If you can help with a decision, that would be great.
> 
> **The problem**
> 
> While working on the `OpenFeature.Providers.DependencyInjection` package, I hit an issue with this configuration:
> 
> ```cs
> builder.Services.AddOpenFeature(of => of
>     .AddProvider(_ => new TestProvider())
>     .AddHook(_ => new TestHook())
> );
> ```
> 
> This fails because:
> 
> * `AddProvider` is an extension for `OpenFeatureProviderBuilder`
> * `AddHook` is an extension for `OpenFeatureBuilder`
> * `AddProvider` returns `OpenFeatureProviderBuilder`, which does **not** have `AddHook`, so the second call is not available and we get a compile-time error.
> 
> If I reverse the order, it works:
> 
> ```cs
> builder.Services.AddOpenFeature(of => of
>     .AddHook(_ => new TestHook())
>     .AddProvider(_ => new TestProvider())
> );
> ```
> 
> Here `AddHook` returns `OpenFeatureBuilder`, which inherits from `OpenFeatureProviderBuilder`, so both methods are available and everything compiles.
> 
> For me this is a design problem, because the API becomes order-dependent and confusing. I tried to make `OpenFeatureProviderBuilder` generic, but that didn’t really solve it.
> 
> I see two options. I’m open to other ideas as well.
> 
> ### Option 1. Move `OpenFeatureBuilder` into `OpenFeature.Providers.DependencyInjection` and remove `OpenFeatureProviderBuilder`
> We would keep all extensions targeting `OpenFeatureBuilder`.
> 
> * **Pros**. Simple. fast to implement.
> * **Cons**. All extensions would depend on the `OpenFeature.Providers.DependencyInjection` package, even when they are not really about providers or DI (for example, hooks).
> 
> ### Option 2. Create `OpenFeature.DependencyInjection.Abstractions` package
> We create a new package. `OpenFeature.DependencyInjection.Abstractions`, move all general DI-related types there, and remove `OpenFeatureProviderBuilder`. All extensions would be built on top of `OpenFeatureBuilder`.
> 
> * **Pros**. Cleaner separation. All DI logic depends on a lightweight abstractions package. In the future, if we want to split hooks or other concerns into separate packages, it will be easier.
> * **Cons**. One more package to maintain. This new package will probably contain most of the logic from the old `OpenFeature.DependencyInjection` package.
> 
> I slightly prefer **Option 2**, even though it introduces an extra package, because it gives us a cleaner and more future-proof structure. cc: @beeme1mr @toddbaert @lukas-reining @thomaspoignant @kinyoklion

Need to have a thought. If this requires another package moving/creation/deprecation we might as well do a v3 like I mentioned in Slack.

I kinda prefer Option B also, but moving the code all the time doesn't bring the necessary stability, even if this is still in Experimental phase.

I would like to know what will be the consequences for the providers and hooks if we go ahead with option B. Which package should it consume? `OpenFeature.Providers.Abstractions` and `OpenFeature.DependencyInjection.Abstractions`?

Then the consumers would just install `OpenFeature`, `OpenFeature.Hosting` and `OpenFeature.Providers.ProviderA`?

To finish, I would believe if we go with option B, we would need to do a v3, make the DI not-experimental and this PR will be part of v3.

arttonoyan committed:
Reorder OpenFeature provider and hook setup

Reordered the configuration of `AddOpenFeature` in `Program.cs` to move `AddInMemoryProvider` after `AddMultiProvider`. This change ensures a specific initialization sequence. Additionally, the order of adding hooks and providers was adjusted to register hooks before providers, potentially affecting execution order and initialization. These changes aim to improve the setup of feature flags and dependency injection.

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

Signed-off-by: Artyom Tonoyan <artonoyan@servicetitan.com>

arttonoyan committed:
Merge branch 'main' into feature/di-abstractions-builder-api-587

askpt commented:
@arttonoyan This PR has several compilation errors. I will have a look when they get fixed. Which option will you think you will implement for the extraction? If there are any necessary breaking changes please rename the title to `feat!: Extract...`

arttonoyan commented:
> @arttonoyan This PR has several compilation errors. I will have a look when they get fixed. Which option will you think you will implement for the extraction? If there are any necessary breaking changes please rename the title to `feat!: Extract...`

@askpt As we discussed earlier, there is a breaking change related to `OpenFeatureBuilder`, so I decided not to fix and merge this PR as is.

Instead, I’ve started working on the next step. extracting `OpenFeature.DependencyInjection.Abstractions`. I created a new branch from this one, and once it’s ready, I’ll open a new PR based on it.

This should properly address the breaking change and also make future work easier. It will let us extract things like handlers and hooks into separate projects without causing more breaking changes.

I’ll open the PR soon.

arttonoyan commented:
I have a follow-up PR https://github.com/arttonoyan/dotnet-sdk/pull/4 that builds on top of this one.

Because PR https://github.com/arttonoyan/dotnet-sdk/pull/4 is based on the branch in my fork (`arttonoyan:feature/di-abstractions-builder-api-587`), I can’t open it directly against `open-feature/dotnet-sdk` until this PR is merged (GitHub requires the PR base branch to exist in the target repo).

What it adds (high level):
- Introduces `OpenFeature.DependencyInjection.Abstractions` and moves the builder/component primitives there
- Updates Hosting + provider DI wiring to use the new split and build-time aggregation

At this stage I’m mainly looking for **architecture feedback** (layering, component approach, package boundaries), not a full nit-level review yet.  
If the community agrees the direction is good, I’ll merge PR https://github.com/arttonoyan/dotnet-sdk/pull/4 into the base branch in my fork, polish it (tests + README/docs), and then open the upstream PR once this PR lands so the code review can continue normally on the SDK side.
@beeme1mr @askpt @kylejuliandev
